src/peergos/shared/storage/CommittableStorage.java

```diff
@@ -0,0 +1,71 @@
+package peergos.shared.storage;
+
+import peergos.shared.crypto.hash.*;
+import peergos.shared.io.ipfs.cid.*;
+import peergos.shared.io.ipfs.multihash.*;
+import peergos.shared.util.*;
+
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.stream.*;
+
+public class CommittableStorage extends DelegatingStorage {
+    private static final int CID_V1 = 1;
+
+    private final ContentAddressedStorage target;
+    private final List<PutArgs> pending = new ArrayList<>();
+
+    public CommittableStorage(ContentAddressedStorage target) {
+        super(target);
+        this.target = target;
+    }
+
+    public synchronized CompletableFuture<Boolean> flush() {
+        List<PutArgs> local = new ArrayList<>(pending);
+        pending.clear();
+        List<CompletableFuture<Multihash>> uploads = local.stream()
+                .map(p -> target.put(p.owner, p.writer, p.signature, p.block, p.tid)
+                        .thenApply(h -> {
+                            if (! h.equals(p.expected))
+                                throw new IllegalStateException("Different hash returned from block write than expected!");
+                            return h;
+                        }))
+                .collect(Collectors.toList());
+        return Futures.combineAllInOrder(uploads)
+                .thenApply(x -> true);
+    }
+
+    @Override
+    public CompletableFuture<Multihash> put(PublicKeyHash owner,
+                                            PublicKeyHash writer,
+                                            byte[] signature,
+                                            byte[] block,
+                                            TransactionId tid) {
+        byte[] sha256 = Arrays.copyOfRange(signature, signature.length - 32, signature.length);
+        Cid cid = hashToCid(sha256, false);
+        synchronized (pending) {
```

Comment:
could use a synchonized List implementation via Collections.synchronizedList.

Reply:
I've reduced the synchronized block. Now it's pretty close to optimal I think.